README.md
# Генератор тестовых COVID-19 QR-кодов

![](demo.png)

Генерируемые тестовые QR-коды подходят для проверки только на тестовом стенде. Данное приложение не
генерирует рабочие QR-коды. Созданными QR-кодами пользоваться запрещено!

Сервис поддерживает дополнительно работу Telegram бота — это опционально, не требуется для работы основного демо-стенда.

## Настройка

#### Основные настройки сайта

Для запуска демо-стенда требуются следующие переменные окружения:

- `FLASK_APP` — указывается местоположение скрипта сервера, необходимо указать: `app/main`
- `FLASK_ENV` — устанавливается значение окружения, требуется для работы в среде разработки, указывается `development`,
по умолчанию ставится `production`.
- `FLASK_DEBUG` — для включения отладочного режима необходимо включить, указав `1`.
- `REDIS_URL` — указывается полный адрес подключения к Redis, например: `redis://localhost:6379/0`

#### Основные настройки Telegram бота

Для работы Telegram бота необходимо предварительно зарегистрировать бота в [@BotFather](https://t.me/BotFather).
Процедуру регистрации можно найти на просторах Интернета.

- `TG_BOT_TOKEN` — токен вашего Telegram бота полученный после регистрации.
- `TG_BOT_QR_HOST` — хост вашего сайта на котором работает основной сайт для тестирования кодов. Поддерживаются IDNA домены.
- `REDIS_URL` — указывается полный адрес подключения к Redis, например: `redis://localhost:6379/0`

### Redis

Для хранения временных данных используется Redis. Все временные данные сохраняются на срок от 10 минут до 1 часа.

### Суперпользователи

По умолчанию задан пользователь `admin` с паролем `password`. Для создания собственных пользователей 
необходимо использовать следующие переменные окружения:

- `ADMIN_USERS` — кол-во суперпользователей.
- `USER_<N>_USERNAME` — логин суперпользователя на сайте.
- `USER_<N>_PASSWORD` — пароль суперпользователя на сайте.
- `USER_<N>_FIRST_NAME` — имя суперпользователя, подставляется автоматически в поле «Имя».
- `USER_<N>_LAST_NAME` — фамилия суперпользователя, подставляется автоматически в поле «Фамилия».
- `USER_<N>_SECOND_NAME` — отчество суперпользователя, подставляется автоматически в поле «Отчество».
- `USER_<N>_B_DAY` — день рождения суперпользователя, подставляется автоматически в поле «День рождения».
- `USER_<N>_SERIES` — первые 2 цифры серии паспорта суперпользователя, подставляется автоматически в поле «Серия паспорта».
- `USER_<N>_NUMBER` — последние 3 цифры номера паспорта суперпользователя, подставляется автоматически в поле «Номер паспорта».
- `USER_<N>_TG_CHAT_ID` — ID пользователя Telegram, используется для автоматической авторизации пользователя в Telegram.
- `USER_<N>_TIMEZONE` — [название часового пояса][tz], используется для автоматической генерации даты и времени валидации тестового QR-кода в указанном часовом поясе пользователя.

Где `<N>` — это цифры от 0 до указанного кол-ва суперпользователей в переменной `ADMIN_USERS`, 0 — это 1-ый пользователь.

[tz]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones#List

## Запуск

Запустить проект можно несколькими способами:

- **[Без использования Docker](#start-without-docker)** — подойдёт для локальной разработки (не рекомендуется).
- **[С частичным использованием Docker](#start-with-part-docker)** — рекомендуемый способ для локальной разработки.
- **[Полностью используя Docker](#start-with-docker)** — подойдёт для запуска в Интернете, для более глубокого тестирования.

<a name="start-without-docker"></a>
### Без использования Docker

Для данного запуска необходимо предварительно позаботиться о предварительно запущенном сервере Redis.

Укажите полный адрес сервера Redis в переменной окружения `REDIS_URL`.

Создайте виртуальное окружение Python 3.9 командой:

```shell
$ python3 -m venv venv
```

Активируйте виртуальное окружение:

```shell
$ source ./venv/bin/active
```

Обновите пакетный менеджер pip до последней версии:

```shell
$ pip insttall --upgrade pip
```

Установите зависимые пакеты:

```shell
$ pip install -r requirements.txt
```

Запустите web-сервер

```shell
$ flask run
```

Для запуска телеграм бота необходимо указать токен вашего бота в переменной `TG_BOT_TOKEN`, после запустить командой:

```shell
$ python app/bot.py
```

<a name="start-with-part-docker"></a>
### С частичным использованием Docker

Этот способ от предыдущего отличается только тем, что Redis можно запустить воспользовавшись Docker. 
Redis с помощью Docker можно запустить 2 способами:

1. С использованием Docker Compose (рекомендуется).
2. Без использования Docker Compose.

Первый вариант делается командой:

```shell
$ docker-compose up redis -d 
```

Второй вариант:

```shell
$ docker run --name redis -d redis -p 6379:6379
```

Далее необходимо выполнить все шаги описанные в предыдущем варианте запуска.

<a name="start-with-docker"></a>
### Полностью используя Docker

Этот способ самый лёгкий для запуска проекта, но требуется заранее установленного Docker и docker-compose.

Для запуска проекта выполните команду:

```shell
$ docker-compose up -d --build
```

## Эксплуатация

### Сайт

Для формирования QR-кода с тестовыми данными перейдите по ссылке: <http://localhost:5000/qr-gen> и авторизуйтесь
под суперпользователем.

Введите тестовые данные и нажмите кнопку «**Создать тестовый QR-код**». Справа появится созданные QR-код с тестовыми
данными. Можно перейти по нему воспользовавшись любым QR сканером.

**ВНИМАНИЕ!** Для тестирования с нескольких устройств необходимо запускать сервис либо на действующем домене, 
либо полностью используя Docker, при этом вместо `localhost` в адресе надо будет указать IP-адрес вашей машины внутри
локальной сети.

### Telegram бот

После запуска соответствующего контейнера (см. инструкцию выше), найдите своего Telegram бота, и нажмите на кнопку 
«Start/Начать» или введите команду `/start`. В ответ вы получите сообщение:

> Username, извиняйте, но вашего ID (<ваш ID в Telegram>) нет в системе.

После этого необходимо добавить этот ID в переменную окружения, для настройки суперпользователя и перезапустить
контейнер с Telegram ботом.

После перезапуска контейнера, снова выполните команду `/start`. Вы должны получить сообщение с данными указанными в 
переменных окружения для суперпользователей с запросом на подтверждение: подтвердите эти данные.

Следующий шаг будет про время валидности QR-кода. Выберите любой из вариантов.

После выбора времени валидности тестового QR-кода вы получите тестовый QR-код с указанием даты и времени до которого 
действителен данный код.


CHANGELOG.md
# Версия 1.1.0

- Добавлено: Поддержка Telegram бота.
- Добавлено: На сайте выводится информация о времени действия тестового QR-кода.

# Версия 1.0.3

- Исправлено: Скрипт сборки проекта, теперь при деплое в Docker Hub верно подставляются версии проекта.

# Версия 1.0.2

- Исправлено: Скрипт сборки проекта. 

# Версия 1.0.1

- Добавлено: Теги с версиями.

# Версия 1.0.0

- Первая стабильная версия. Ура!


app/tests.py
import os
import random
import string
from datetime import timedelta

import unittest
import main as tested_app


class FlaskAppTest(unittest.TestCase):
    def setUp(self) -> None:
        tested_app.app.config['TESTING'] = True
        self.app = tested_app.app.test_client()
        setattr(self.app, 'users', tested_app.users)
        self.redis_client = tested_app.redis_client

    def test_get_home(self):
        r = self.app.get('/')
        self.assertEqual(r.status_code, 302)

    def test_get_covid_cert_verify_fail(self):
        unrz = self.get_unrz()
        r = self.app.get(f'/covid-cert/verify/{unrz}')
        self.assertEqual(r.status_code, 302)

    def test_get_covid_cert_verify(self):
        unrz = self.get_unrz(write_to_redis=True)
        r = self.app.get(f'/covid-cert/verify/{unrz}')
        self.assertEqual(r.status_code, 200)

    def test_get_covid_confid(self):
        r = self.app.get('/covid-web/config.json')

        self.assertEqual(r.status_code, 200)
        self.assertEqual(r.is_json, True)

        data = r.json
        self.assertEqual(data['covidCertCheckUrl'], '/api/covid-cert/v3/cert/check/')

    def test_get_covid_cert_check_fail(self):
        unrz = self.get_unrz()
        r = self.app.get(f'/api/covid-cert/v3/cert/check/{unrz}')
        self.assertEqual(r.status_code, 200)
        self.assertEqual(r.is_json, True)
        self.assertEqual(r.json, {})

    def test_get_covid_cert_check(self):
        unrz = self.get_unrz(write_to_redis=True)
        r = self.app.get(f'/api/covid-cert/v3/cert/check/{unrz}')
        self.assertEqual(r.status_code, 200)
        self.assertEqual(r.is_json, True)

        data = r.json['items'][0]
        self.assertEqual(data['unrzFull'], unrz)
        self.assertEqual(data['qr'], '')

    def get_unrz(self, write_to_redis=False):
        unrz = ''.join(random.choice(string.digits) for _ in range(16))

        if write_to_redis:
            data: dict = self.app.users['admin']
            data['qr'] = ''
            for key, value in data.items():
                self.redis_client.hset(unrz, key, value)
            self.redis_client.expire(unrz, timedelta(seconds=10))

        return unrz


if __name__ == '__main__':
    unittest.main()


app/bot.py
import base64
import io
import os
import random
import string
import uuid
from datetime import timedelta, datetime
from urllib.parse import urlparse

import idna
import pytz
import redis
from pyqrcode import QRCode
from telegram import Update, InlineKeyboardMarkup, InlineKeyboardButton
from telegram.ext import Updater, CommandHandler, CallbackContext, CallbackQueryHandler

redis_url = os.getenv('REDIS_URL', 'redis://localhost:6379/0')
redis_url = urlparse(redis_url)

redis_client = redis.Redis(host=redis_url.hostname, port=redis_url.port, db=redis_url.path[1:])

users = {}
if os.getenv('ADMIN_USERS') and int(os.getenv('ADMIN_USERS')) > 0:
    for idx in range(int(os.getenv('ADMIN_USERS'))):
        users[os.getenv(f'USER_{idx}_TG_CHAT_ID')] = {
            'first_name': os.getenv(f'USER_{idx}_FIRST_NAME'),
            'last_name': os.getenv(f'USER_{idx}_LAST_NAME'),
            'second_name': os.getenv(f'USER_{idx}_SECOND_NAME'),
            'b_day': os.getenv(f'USER_{idx}_B_DAY'),
            'series': os.getenv(f'USER_{idx}_SERIES'),
            'number': os.getenv(f'USER_{idx}_NUMBER'),
            'timezone': os.getenv(f'USER_{idx}_TIMEZONE'),
        }


def start(update: Update, context: CallbackContext) -> None:
    if str(update.effective_chat.id) in users:
        user = users[str(update.effective_chat.id)]
        keyword = [
            [
                InlineKeyboardButton('Всё верно!', callback_data='passport_yes'),
                InlineKeyboardButton('Не то!', callback_data='passport_no')
            ]
        ]
        reply_markup = InlineKeyboardMarkup(keyword)
        update.message.reply_html(
            (
                'Подтвердите правильность данных:\n\n'
                'ФИО: <b>{} {} {}</b>\n'
                'День рождения: <b>{}</b>\n'
                'Серия и номер паспорта: <b>{}** ***{}</b>'.format(
                    user['last_name'], user['first_name'], user['second_name'],
                    user['b_day'],
                    user['series'], user['number']
                )
            ),
            reply_markup=reply_markup
        )
    else:
        update.message.reply_text(
            f'{update.effective_user.first_name}, извиняйте, но вашего ID ({update.effective_chat.id}) нет в системе.'
        )


def button(update: Update, context: CallbackContext) -> None:
    query = update.callback_query
    query.answer()
    if query.data == 'passport_yes':
        keyboard = [
            [InlineKeyboardButton('1 час', callback_data='3600')],
            [InlineKeyboardButton('30 минут', callback_data='1800')],
            [InlineKeyboardButton('10 минут', callback_data='600')],
        ]
        reply_markup = InlineKeyboardMarkup(keyboard)
        query.edit_message_text('На какой период времени создать QR-код?', reply_markup=reply_markup)
    elif query.data == 'passport_no':
        query.edit_message_text('Пока нет возможности исправлять данные паспорта через Telegram бота.')
    elif query.data in ['3600', '1800', '600']:
        url = os.getenv('TG_BOT_QR_HOST')
        try:
            url = idna.encode(url).decode('utf-8')
        except idna.core.InvalidCodepoint:
            pass
        unrz = ''.join(random.choice(string.digits) for _ in range(16))
        ck = uuid.uuid4().hex.replace('-', '')
        url = 'https://{}/covid-cert/verify/{}?lang=ru&ck={}'.format(
            url,
            unrz,
            ck
        )
        qr = QRCode(url)
        media = io.BytesIO()
        qr.png(media, scale=4)
        user = users[str(query.message.chat_id)]
        qr_code = media.getvalue()
        qr_config = user
        qr_config['qr'] = base64.b64encode(qr_code).decode('utf-8')
        for key, value in qr_config.items():
            redis_client.hset(unrz, key, value)
        redis_client.expire(unrz, timedelta(seconds=int(query.data)))
        user_datetime = datetime.now(pytz.timezone(user['timezone'])) + timedelta(seconds=int(query.data))
        caption = 'Этот QR-код надо сохранить в фото.\n\nКод действителен до: {}'.format(
            user_datetime.strftime('%H:%M:%S %d.%m.%Y (%Z)')
        )
        query.delete_message()
        query.message.reply_photo(photo=qr_code, caption=caption)


def main() -> None:
    updater = Updater(os.getenv('TG_BOT_TOKEN'))

    updater.dispatcher.add_handler(CommandHandler('start', start))
    updater.dispatcher.add_handler(CallbackQueryHandler(button))

    updater.start_polling()
    updater.idle()


if __name__ == '__main__':
    main()


app/main.py
import base64
import io
import os
import random
import secrets
import string
import uuid
from datetime import timedelta, datetime
from urllib.parse import urljoin

import idna
import pytz
from dateutil.relativedelta import relativedelta
from fakeredis import FakeRedis
from flask import Flask, request, send_file, render_template, redirect, url_for
from flask_httpauth import HTTPDigestAuth
from flask_redis import FlaskRedis
from pyqrcode import QRCode
from transliterate import translit

app = Flask(__name__)
app.config['SECRET_KEY'] = secrets.token_hex(32)
app.config['REDIS_URL'] = os.getenv('REDIS_URL', 'redis://localhost:6379/0')
auth = HTTPDigestAuth()
if app.testing:
    redis_client = FakeRedis(app)
else:
    redis_client = FlaskRedis(app)

users = {
    'admin': {
        'password': 'password',
        'first_name': 'Иван',
        'last_name': 'Иванов',
        'second_name': 'Иванович',
        'b_day': '2020-01-01',
        'series': '12',
        'number': '789',
        'tz': 'Europe/Moscow'
    }
}
if os.getenv('ADMIN_USERS', 0) and int(os.getenv('ADMIN_USERS', 0)) > 0:
    users = {}
    for idx in range(int(os.getenv('ADMIN_USERS'))):
        users[os.getenv(f'USER_{idx}_USERNAME')] = {
            'password': os.getenv(f'USER_{idx}_PASSWORD'),
            'first_name': os.getenv(f'USER_{idx}_FIRST_NAME'),
            'last_name': os.getenv(f'USER_{idx}_LAST_NAME'),
            'second_name': os.getenv(f'USER_{idx}_SECOND_NAME'),
            'b_day': os.getenv(f'USER_{idx}_B_DAY'),
            'series': os.getenv(f'USER_{idx}_SERIES'),
            'number': os.getenv(f'USER_{idx}_NUMBER'),
            'tz': os.getenv(f'USER_{idx}_TIMEZONE')
        }
setattr(app, 'users', users)


@auth.get_password
def get_pw(username):
    if username in users:
        return users[username]['password']
    return None


@app.route('/')
def home():
    return redirect('https://www.gosuslugi.ru', code=302)


@app.route('/covid-cert/verify/<string:unrz>')
def covid_cert_verify(unrz: str):
    if not redis_client.exists(unrz):
        query = request.query_string.decode('utf-8').replace("'", '')
        real_url_check = f'https://www.gosuslugi.ru/covid-cert/verify/{unrz}?{query}'
        return redirect(real_url_check)
    return render_template('covid-cert.html', unrz=unrz)


@app.route('/covid-web/config.json')
def covid_config():
    return {
        "production": True,
        "baseUrl": "https://www.gosuslugi.ru/",
        "betaUrl": "https://www.gosuslugi.ru/",
        "lkUrl": "https://lk.gosuslugi.ru/",
        "lkApiUrl": "//www.gosuslugi.ru/api/lk/v1/",
        "yaCounter": 0,
        "authProviderUrl": "//www.gosuslugi.ru/auth-provider/login?rUrl=",
        "nsiApiUrl": "//www.gosuslugi.ru/api/nsi/v1/",
        "staticDomainLibAssetsPath": "//gu-st.ru/covid-web-st/lib-assets/",
        "timingApiUrl": "//www.gosuslugi.ru/health",
        "staticDomainAssetsPath": "//gu-st.ru/covid-web-st/assets/",
        "appStores": {
            "appStore": "https://redirect.appmetrica.yandex.com/serve/529060629282032912",
            "googlePlay": "https://redirect.appmetrica.yandex.com/serve/745233407570662167",
            "appGallery": "https://appgallery8.huawei.com/#/app/C101280309"
        },
        "socialNetworks": {
            "vk": "https://vk.me/new.gosuslugi",
            "ok": "https://ok.ru/gosuslugi",
            "fb": "https://www.facebook.com/new.gosuslugi",
            "tg": "https://t.me/gosuslugi"
        },
        "portalCfgUrl": "//www.gosuslugi.ru/api/portal-cfg/",
        "mainBlocksData": "//www.gosuslugi.ru/api/mainpage/v4",
        # "covidCertCheckUrl": "//www.gosuslugi.ru/api/covid-cert/v3/cert/check/",
        "covidCertCheckUrl": "/api/covid-cert/v3/cert/check/",
        "covidCertUrl": "//www.gosuslugi.ru/api/covid-cert/v2/",
        "registerCovidUrl": "//www.gosuslugi.ru/api/register-covid/v2/",
        "vaccineUrl": "//www.gosuslugi.ru/api/vaccine/v1/",
        "covidCertPdfUrl": "//www.gosuslugi.ru/api/covid-cert/v1/cert/{unrzFull}/pgu/srfile/pdf",
        "vaccineUrlv2": "//www.gosuslugi.ru/api/vaccine/v2/",
        "quadrupelUrl": "//www.gosuslugi.ru/api/quadrupel/v1/",
        "grpIdCheck": ["RA.USR_CFM", "RA_TOOL_SUPERACCESS"]
    }


@app.route('/api/covid-cert/v3/cert/check/<string:unrz>')
def covid_cert_check(unrz: str):
    if not redis_client.exists(unrz):
        return {}

    config = redis_client.hgetall(unrz)
    config = {x.decode('utf-8'): config.get(x).decode('utf-8') for x in config}

    en_first_name = translit(config['first_name'], 'ru', reversed=True)
    en_last_name = translit(config['last_name'], 'ru', reversed=True)
    en_second_name = translit(config['second_name'], 'ru', reversed=True)

    fio = [
        '{} {} {}'.format(
            config['last_name'][0] + '*' * (len(config['last_name']) - 1),
            config['first_name'][0] + '*' * (len(config['first_name']) - 1),
            config['second_name'][0] + '*' * (len(config['second_name']) - 1),
        ),
        '{} {} {}'.format(
            en_last_name[0] + '*' * (len(en_last_name) - 1),
            en_first_name[0] + '*' * (len(en_first_name) - 1),
            en_second_name[0] + '*' * (len(en_second_name) - 1),
        ),
    ]

    b_day = datetime.strptime(config['b_day'], '%Y-%m-%d')

    recovery_date = datetime.now() - relativedelta(months=2)
    valid_until = recovery_date + relativedelta(months=6)

    return dict({
        'items': [
            {
                'type': 'ILLNESS_FACT',
                'unrz': '123123132',
                'unrzFull': unrz,
                'attrs': [
                    {
                        'type': 'date',
                        'title': 'Дата выздоровления',
                        'entitle': 'Recovery date',
                        'envalue': recovery_date.strftime('%d.%m.%Y'),
                        'value': recovery_date.strftime('%d.%m.%Y'),
                        'order': 1,
                    },
                    {
                        'type': 'date',
                        'title': 'Действует до',
                        'entitle': 'Valid until',
                        'envalue': valid_until.strftime('%d.%m.%Y'),
                        'value': valid_until.strftime('%d.%m.%Y'),
                        'order': 1,
                    },
                    {
                        'type': 'fio',
                        'title': 'ФИО',
                        'entitle': 'Full name',
                        'envalue': fio[1],
                        'value': fio[0],
                        'order': 3,
                    },
                    {
                        'type': 'passport',
                        'title': 'Паспорт',
                        'entitle': 'National passport',
                        'envalue': '{}** ***{}'.format(config['series'], config['number']),
                        'value': '{}** ***{}'.format(config['series'], config['number']),
                        'order': 4,
                    },
                    {
                        'type': 'birthDate',
                        'title': 'Дата рождения',
                        'entitle': 'Date of birth',
                        'envalue': b_day.strftime('%d.%m.%Y'),
                        'value': b_day.strftime('%d.%m.%Y'),
                        'order': 6,
                    },
                ],
                'title': 'Сведения о перенесенном заболевании COVID-19',
                'entitle': 'Previous disease COVID-19',
                'qr': config['qr'],
                'status': '1',
                'order': 0,
                'expiredAt': valid_until.strftime('%d.%m.%Y'),
                'serviceUnavailable': False,
            }
        ],
        'hasNext': False,
    })


@app.route('/qr-gen', methods=['GET', 'POST'])
@auth.login_required
def qr_generator():
    qr_code = None
    url = None
    user = users[auth.current_user()]
    ttl = datetime.now()

    if request.method == 'POST':
        # Генерируем случайный код пациента
        unrz = ''.join(random.choice(string.digits) for x in range(16))
        # Генерируем случайный код проверки
        ck = uuid.uuid4().hex.replace('-', '')
        # Указываем язык по умолчанию
        lang = 'ru'
        # Создает query параметры для URL
        query_params = f'{lang=}&{ck=}'
        # Создаём URL для будущего QR-кода
        host = idna.encode(request.host)
        host = 'https://{}/'.format(host.decode('utf-8'))
        url = urljoin(host, url_for('covid_cert_verify', unrz=unrz))
        url += '?' + query_params.replace("'", '')

        # Создаём необходимый нам QR-код
        qr = QRCode(url)
        buffer = io.BytesIO()
        qr.png(buffer, scale=4)
        qr_code = base64.b64encode(buffer.getvalue()).decode('utf-8')

        # Получаем данные из формы и записываем в Redis с указанным сроком годности
        qr_config = request.form.to_dict()
        qr_config['qr'] = qr_code
        for key, value in qr_config.items():
            redis_client.hset(unrz, key, value)
        redis_client.expire(unrz, timedelta(seconds=int(qr_config.get('expire', 3600))))
        ttl = datetime.now(pytz.timezone(user['tz'])) + timedelta(seconds=int(qr_config.get('expire', 3600)))

    return render_template('qr-generator.html',
                           username=auth.current_user(),
                           user=user,
                           ttl=ttl.strftime('%H:%M:%S %d.%m.%Y (%Z)'),
                           url=url, qr_code=qr_code)


@app.route('/<path:path>')
def route_frontend(path):
    file_path = os.path.join(app.static_folder, path)
    if os.path.isfile(file_path):
        return send_file(file_path)
    else:
        return redirect('https://www.gosuslugi.ru')


if __name__ == '__main__':
    app.run()
